### PR TITLE
PushPX: GPU kernel optimization

### PR DESCRIPTION
**Pull request by WeiqunZhang**
The GatherAndPush kernel in the PushPX function has a very low occupancy due to register pressure.  There are a number of reasons.  By default, we compile with QED module on, even if we do not use it at run time.  Another culprit is the GetExternalEB functor that contains 7 Parsers.  Again, we have to pay a high runtime cost, even if we do not use it.  In this PR, we move some runtime logic out of the GPU kernel to eleminate the unnecessary cost if QED and GetExternalEB are not used at run time.  This uses a fold expression trick that @AlexanderSinn taught us.

Here are some performance results before this PR.

    | QED | GetExternalEB | Time |
    |-----+---------------+------|
    | On  | On            | 2.17 |
    | Off | On            | 1.79 |
    | Off | Commented out | 1.34 |

Note that in the tests neither QED nor GetExternalEB is actually used at run time.  But the extra cost is very high.  With this PR, the kernel time is the same as that when both QED and GetExternalEB are disabled at compile time, even though both options are disabled at run time.

More information on the kernels compiled for MI250X.  The most expensive variant with both QED and GetExternalEB on has

    NumSgprs: 108
    NumVgprs: 256
    NumAgprs: 40
    TotalNumVgprs: 296
    ScratchSize: 264
    Occupancy: 1

The cheapest variant with both QED and GetExternalEB disabled has

    NumSgprs: 104
    NumVgprs: 249
    NumAgprs: 0
    TotalNumVgprs: 249
    ScratchSize: 144
    Occupancy: 2

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
It works with HIP.  Let's see if it compiles with CUDA and SYCL.

File: Source/Particles/PhysicalParticleContainer.cpp [hunk 1]
```diff
@@ -109,6 +109,34 @@
...
+                  amrex::ParallelFor(N, [f] AMREX_GPU_DEVICE (Long i) {
+                      f(i, CTOption<Op>{});
+                  })
+              )
+            : (
+                ++option_miss, void()
+              )
+        ),...
+    );
+    AMREX_ASSERT(option_miss < sizeof...(option_pack));
```
Comment (AlexanderSinn) [inline Source/Particles/PhysicalParticleContainer.cpp]:
```suggestion
    AMREX_ASSERT(option_miss + 1 == sizeof...(Op));
```
Should be a bit safer this way

**Comment by WeiqunZhang** [comment]
Unfortunately, it does not work with nvcc >= 15.
```
An extended __device__ lambda cannot first-capture variable in constexpr-if context.
```

**Comment by AlexanderSinn** [comment]
I had that problem a few times before, the default lambda capture `[=]` needs to be replaced with one explicitly mentioning all of the needed variables, including the one only used inside the constexpr-if.

https://github.com/Hi-PACE/hipace/blob/98c62be645c8a42694b9c75c880da1ba1fe5bdb1/src/particles/deposition/PlasmaDepositCurrentInner.H#L188-L201

**Comment by WeiqunZhang** [comment]
The problem is it might be very expensive to capture something that is not needed in the case of `GetExternalEB`.

**Comment by AlexanderSinn** [comment]
Lambda capture is a fundamental C++ inefficiency and I think there is no way to reduce that without adding a wrapper around PushPX, which calls one multiple PushPX templates.
I had a setup to somewhat fix this here: https://github.com/Hi-PACE/hipace/pull/775/commits/470f6ed1be9a90a00241d9c0ec775838d7d71dc8 but removed it because of the negligible performance impact in that case.
The hope is that the GPU code doesn’t access all of the captured data, maybe capturing pointers to data in GPU memory instead would help.

**Comment by WeiqunZhang** [comment]
After working about the lambda capture issue, we now have a redefinition problem with the new nvcc compiler.  :disappointed:   I will try to do this in yet another way.

**Comment by WeiqunZhang** [comment]
Yet another attempt #3402

**Pull request closed without merging**